src/imageattributes.c

```diff
@@ -204,23 +218,98 @@ gdip_process_bitmap_attributes (GpBitmap *bitmap, void **dest, GpImageAttributes
 
 	}
 
+	/* Treshold correction */
+	if (!(treshold->flags & ImageAttributeFlagsNoOp) && treshold->flags & ImageAttributeFlagsThresholdEnabled) {
+		BYTE cutoff = (BYTE)round(treshold->threshold * 255.0);
+
+		for (int y = 0; y < bitmap->active_bitmap->height; y++) {
+			for (int x = 0; x < bitmap->active_bitmap->width; x++) {
+
+				BYTE r, g, b, a;
+
+				GdipBitmapGetPixel(bmpdest, x, y, &color);
+
+				a = (color & 0xff000000) >> 24;
+				r = (color & 0x00ff0000) >> 16;
+				g = (color & 0x0000ff00) >> 8;
+				b = (color & 0x000000ff);
+
+				r = r > cutoff ? 255 : 0;
+				g = g > cutoff ? 255 : 0;
+				b = b > cutoff ? 255 : 0;
+
+				color = ((guint32)a << 24) | (r << 16) | (g << 8) | b;
+				GdipBitmapSetPixel(bmpdest, x, y, color);
+			}
+		}
+	}
+
+	/* CMYK calculation */
+	if (!(cmyk->flags & ImageAttributeFlagsNoOp) && cmyk->flags & ImageAttributeFlagsOutputChannelEnabled) {
+		for (int y = 0; y < bitmap->active_bitmap->height; y++) {
+			for (int x = 0; x < bitmap->active_bitmap->width; x++) {
+
+				BYTE r, g, b, a, C, M, Y, K;
+
+				GdipBitmapGetPixel(bmpdest, x, y, &color);
+
+				a = (color & 0xff000000) >> 24;
+				r = (color & 0x00ff0000) >> 16;
+				g = (color & 0x0000ff00) >> 8;
+				b = (color & 0x000000ff);
+
+				C = 255 - r;
+				M = 255 - g;
+				Y = 255 - b;
+				K = min(min(C, M), Y);
+				
+				/* correct complementary color lever based on k */
+				C -= K;
+				M -= K;
+				Y -= K;
+
+				switch (cmyk->outputchannel_flags) {
+				case ColorChannelFlagsC:
+					r = g = b = C;
+					break;
+				case ColorChannelFlagsM:
+					r = g = b = M;
+					break;
+				case ColorChannelFlagsY:
+					r = g = b = Y;
+					break;
+				case ColorChannelFlagsK:
+					r = g = b = K;
+					break;
+				}
+
+				color = ((guint32)a << 24) | (r << 16) | (g << 8) | b;
+				GdipBitmapSetPixel(bmpdest, x, y, color);
+			}
+		}
+	}
+
 	/* Apply transparency range */
-	if (trans->flags & ImageAttributeFlagsColorKeysEnabled) {
+	/* FIXME: This will convert pixels which have a color in [key_colorlow, key_colorhey] to transparent pixels.
```

Comment:
@filipnavara Any thoughts on how ti fix this? The code correctly sets the pixels to be transparent, but `cairo_fill` will not copy transparent pixels, effectively rendering this a no-op.

Reply:
@qmfrederik I briefly looked at it but it will require me to dive deeper into the code. I will review the remaining patches and try to get back to this one next week.